rust/codelist-builder-rs/src/snomed_usage_data.rs

```diff
@@ -0,0 +1,391 @@
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct SnomedUsageDataEntry {
+    pub snomed_concept_id: String,
+    pub description: String,
+    pub usage: String, // allows for * for count of 1-4
+    pub active_at_start: bool,
+    pub active_at_end: bool,
+}
+
+/// Struct to represent snomed usage data
+///
+/// # Fields
+/// * `usage_data` - The usage data
+/// * `usage_year` - The usage year
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct SnomedUsageData {
+    pub usage_data: Vec<SnomedUsageDataEntry>,
+    pub usage_year: UsageYear,
+}
+
+impl SnomedUsageData {
+    /// Download snomed usage data from a url
+    ///
+    /// # Arguments
+    /// * `base_url` - The base url
+    /// * `usage_year` - The usage year
+    ///
+    /// # Returns
+    /// Self or an error if the download fails
+    pub async fn download_usage(
+        base_url: &str,
+        usage_year: UsageYear,
+    ) -> Result<Self, CodeListBuilderError> {
+        let url = format!(
+            "{}/{}",
+            base_url.trim_end_matches('/'),
+            usage_year.path().trim_start_matches('/')
+        );
+
+        let body = reqwest::get(&url).await?.text().await?;
+
+        let usage_data = Self::parse_from_string(&body)?;
+
+        Ok(SnomedUsageData { usage_data, usage_year })
+    }
+
+    /// Parse snomed usage data from a string
+    ///
+    /// # Arguments
+    /// * `data` - The data to parse
+    ///
+    /// # Returns
+    /// * The parsed usage data or an error
+    pub fn parse_from_string(
+        data: &str,
+    ) -> Result<Vec<SnomedUsageDataEntry>, CodeListBuilderError> {
+        let mut rdr = csv::ReaderBuilder::new()
+            .has_headers(true)
+            .delimiter(b'\t')
+            .from_reader(data.as_bytes());
+
+        let mut usage_data = Vec::new();
+
+        for (row_idx, result) in rdr.records().enumerate() {
+            let record = result?;
+
+            if record.len() != 5 {
+                return Err(CodeListBuilderError::invalid_usage_data(format!(
+                    "Invalid number of columns in record ({}) at row {}",
+                    record.len(),
+                    row_idx + 1
+                )));
+            }
+
+            if let Some((col_idx, _)) =
+                record.iter().enumerate().find(|(_, field)| field.trim().is_empty())
+            {
+                return Err(CodeListBuilderError::invalid_usage_data(format!(
+                    "Empty value found in record at row {}, column {}",
+                    row_idx + 1,
+                    col_idx
+                )));
+            }
+
+            let entry = SnomedUsageDataEntry {
+                snomed_concept_id: record[0].to_string(),
+                description: record[1].to_string(),
+                usage: record[2].to_string(),
+                active_at_start: record[3] == *"1",
+                active_at_end: record[4] == *"1",
```

Comment:
Can you add a comment what this is doing? it won;t be obvious to a non expert that `1` is being used to denotes something is active. We should put an explainer about this in the readme or at the top of this file where we explain the "rules" of the file.